[RDY] Rewrite English man pages in mdoc, update for Nvim

Pyrohh opened the pull request:
continues #2395, refs #1826

You don't need to know mdoc to review this; just open up `nvim.1` and/or `nvimtutor.1` with `man -l`. Alternatively you can locally view the PDF files I've generated using mandoc (it can also output HTML for the website). I'll remove them before this is merged

As I said there's no need to, but if you want to learn mdoc then I recommend this guide: http://manpages.bsd.lv/
It's very comprehensive, but mdoc is quite simple so it only took a few hours.

---

@splinterofchaos I'm not sure I addressed what you said earlier (https://github.com/neovim/neovim/pull/2395#discussion_r28200533) as you intended, so please take a look.

edit: oh wow, you can apparently preview PDFs on GitHub (the text rendering isn't great though): https://github.com/Pyrohh/neovim/blob/manpages/man/nvim.pdf

Pyrohh commented on man/nvim.1:
```diff
@@ -0,0 +1,375 @@
+.El
+.Sh ENVIRONMENT
+.Bl -tag -width VIMRUNTIME
+.It Ev VIMINIT
```
Some of these env variable descriptions aren't very good, probably because I don't really know what differentiates some of them. Does anyone have any improvements?

kopischke replied:
`VIMINIT` is only used outside Vim: it contains commands to execute on Vim startup (as a literal ex command line). For `VIMRUNTIME` and `VIM`, the best references are `help $VIMRUNTIME` and `help $VIM` (in a nutshell, `VIMRUNTIME` is the application provided runtime path, while `VIM` is probably best seen a way to set a shared runtime path on a system).

Pyrohh replied:
> VIMINIT is only used outside Vim: it contains commands to execute on Vim startup (as a literal ex command line)

:+1: thanks

> (in a nutshell, VIMRUNTIME is the application provided runtime path, while VIM is probably best seen a way to set a shared runtime path on a system)

thanks, I'll look into that.

Pyrohh replied:
I've updated this by the way, in case you want to take a look.

splinterofchaos commented on man/nvim.1:
```diff
@@ -0,0 +1,375 @@
+commands.
+.It Fl
+Read text from standard input until EOF, then open a buffer with that text.
+Commands are read from standard error, which should be a terminal.
```
I think TTY or PTY would be more accurate, but this sounds odd because an output stream isn't a terminal. Maybe "should connect to a terminal".

Pyrohh replied:
> I think TTY or PTY would be more accurate

I used terminal because it's more generic, and TTY stands for teletype ([among other things apparently](https://en.wikipedia.org/wiki/Teleprinter)), which is probably wrong in most modern contexts.

> but this sounds odd because an output stream isn't a terminal. Maybe "should connect to a terminal".

:+1: perhaps "... , which should be associated to a terminal."?

splinterofchaos replied:
:+1:

splinterofchaos commented on man/nvim.1:
```diff
@@ -0,0 +1,375 @@
+Interpret filenames literally, that is do not expand wildcards.
+Has no effect on UNIX-like systems, where the shell expands wildcards.
+.It Fl e
+Ex mode.
```
"Start in Ex mode"?

Pyrohh replied:
:+1: I'll update the rest of the "mode" flags, but keep the `--help` output the same for brevity.

Pyrohh replied:
Eh, I'm not sure actually, as this introduces a lot of redundancy.

fmoralesc replied:
Also, add `see :help Ex-mode`.

splinterofchaos commented on man/nvim.1:
```diff
@@ -0,0 +1,375 @@
+.Fl s .
+.It Fl d
+Diff mode.
+Shows the difference between up to four files.
```
We don't give any example of using `-d`. Maybe just putting `[files...]` after the `-d` would suffice.

Pyrohh replied:
:+1:

splinterofchaos commented:
> @splinterofchaos I'm not sure I addressed what you said earlier (#2395 (comment)) as you intended, so please take a look.

Yeah, it's the `+[linenum]`, `+/pattern`, and `-c command`. `[linenum]` and `/pat` are both commands and one could equivalently do `-c [linenum]`, `-c /pat`, or `+command`.

Pyrohh commented:
Repushed with suggested changes (and some other ones) and regenerated the PDF.

@splinterofchaos 
> Yeah, it's the +[linenum], +/pattern, and -c command. [linenum] and /pat are both commands and one could equivalently do -c [linenum], -c /pat, or +command.

Ah, I understand. I mentioned that in addition to what was already there in https://github.com/Pyrohh/neovim/commit/20b3e79801041f15dff06508b5cd64d4396db602.

fmoralesc commented on man/nvim.1:
```diff
@@ -0,0 +1,388 @@
+.Ar nvimrc
+is
+.Cm NONE ,
+do not read any initialization files or environment variables.
```
You need to mention `-u NORC` too, I think, which doesn't load the nvimrc file, but loads system-wide plugins.

Pyrohh replied:
:+1:

fmoralesc commented on man/nvim.1:
```diff
@@ -0,0 +1,388 @@
+.It Fl e
+Ex mode.
+.It Fl E
+Improved Ex mode.
```
Add `see :help gQ`

Pyrohh replied:
:+1:

fmoralesc commented on man/nvim.1:
```diff
@@ -0,0 +1,388 @@
+.It Fl s
+Silent mode.
+Only takes effect if
+.Fl e
```
It also takes effect if `-E` is specified.

Pyrohh replied:
:+1:

fmoralesc commented on man/nvim.1:
```diff
@@ -0,0 +1,388 @@
+Implied by
+.Cm -u NONE .
+.It Fl o Ns Op Ar N
+Open
```
Neither this nor the help say what happens when N is less than the number of file arguments given (it gives the first N files windows or tabpages and hides the rest). Just as a heads up.

fmoralesc replied:
Same for `-O` and `-p`.

Pyrohh replied:
:+1:

fmoralesc commented on man/nvim.1:
```diff
@@ -0,0 +1,388 @@
+can be used.
+.It Fl S Op Ar session
+Source
+.Ar session
```
What is a session? Maybe add `see :help session-file`

Pyrohh replied:
:+1:

fmoralesc commented on man/nvim.1:
```diff
@@ -0,0 +1,388 @@
+.Ar session
+is omitted, then
+.Pa Session.vim
+is used (only works if
```
"is looked for and, if found, used" (phrase it better than me, obvs, though).

Pyrohh replied:
:+1:

Pyrohh commented:
@fmoralesc Thanks for the review, I pushed your changes. I'll squash everything and mark as RDY if it looks good to you (and everyone else).

fmoralesc commented:
LGTM, with my last comment regarding `NORC`.

Pyrohh commented:
I attempted to address your comment @fmoralesc.

Pyrohh commented:
I set this to RDY; I'll wait a day or two to merge if there's no further comments.

justinmk commented on src/nvim/main.c:
```diff
@@ -905,12 +905,7 @@ static void command_line_scan(mparm_T *parmp)
           if (STRICMP(argv[0] + argv_idx, "help") == 0)
             usage();
           else if (STRICMP(argv[0] + argv_idx, "version") == 0) {
-            Columns = 80;                 /* need to init Columns */
```
why doesn't Columns need to be initialized?

Pyrohh replied:
It's already done in `globals.h`.

justinmk commented on src/nvim/main.c:
```diff
@@ -905,12 +905,7 @@ static void command_line_scan(mparm_T *parmp)
-            msg_putchar('\n');
-            msg_didout = FALSE;
-            mch_exit(0);
+            version();
```
It would be better to leave the explicit mch_exit(0) in both cases, instead of hiding it in version() and having to explain what's going on.

Pyrohh replied:
:+1:

justinmk commented on src/nvim/main.c:
```diff
@@ -1985,41 +1992,42 @@ static void usage(void)
   mch_msg(_("  +                     Start at end of file\n"));
-  mch_msg(_("  +<lnum>               Start at line <lnum>\n"));
+  mch_msg(_("  +<linenum>            Start at line <linenum>\n"));
+  mch_msg(_("  +/<pattern>           Start at first occurrence of <pattern>\n"));
```
:)

Pyrohh replied:
Heh, I wasn't even aware it was missing until I checked the man page :grin:

justinmk commented on runtime/doc/Makefile:
```diff
@@ -51,12 +30,6 @@ tags: doctags $(DOCS)
-
-# OSX groff doesn't support utf-8 as input encoding, so this won't work there.
-.1.man:
-	groff -k -mandoc -Tutf8 $< | sed -e s/.^H//g > $@
```
why was this removed?

Pyrohh replied:
After seeing this[1] I wasn't sure that it's actually true. Assuming a moderately recent (last decade) version of groff, man pages written in mdoc don't need any pre-processing; you just type `man my-program` and your man page is formatted on the fly, then displayed.

Could you test how `nvim.1` and `nvimtutor.1` look on OS X? You should be able to view them using `man -l`.

[1]: https://developer.apple.com/library/mac/documentation/Darwin/Reference/ManPages/man1/groff.1.html

justinmk replied:
It looks fantastic actually.

On Mac (BSD I guess), there's no `man -l`, but a slash in the path works similarly: `man ./nvim.1`.

Pyrohh replied:
Thanks for checking, that's great to hear (I was afraid we'd have to do some magic to it).

justinmk commented:
So the reason the manpages were moved out of `runtime/doc/`, is that we don't want to ship them, just install them? I am not an advanced manpage user, but I wonder if anyone objects to that. Is there any reason a user would want the `.1` files in their `/user/share/...`?

Pyrohh commented:
> So the reason the manpages were moved out of runtime/doc/, is that we don't want to ship them, just install them?

Yes, pretty much. I don't think it makes sense them being in runtime, because... well they're not part of the nvim runtime.

> Is there any reason a user would want the .1 files in their /user/share/...?

If the man pages are not installed to `/usr/{share}/man/man[1-8]`, or wherever a particular systems manual viewer expects them (could be specified via `man.conf`, `MANPATH`, etc.), then typing `man nvim` won't find the manual page.

Pyrohh commented:
@justinmk Thanks for taking a look, repushed with your suggestions, and fixed up that commit message.

justinmk commented on man/nvim.1:
```diff
@@ -0,0 +1,417 @@
+.It Fl -api-info
+Dump API metadata serialized to msgpack and exit.
+.It Fl -embed
+Use standard input and standard output as a msgpack-rpc channel.
```
stdin and stdout

Pyrohh replied:
I had it that way before, but changed it since they aren't being talked about in programming contexts (i.e., the actual file descriptors).

justinmk commented on man/nvim.1:
```diff
@@ -0,0 +1,417 @@
+commands.
+.It Fl
+Read text from standard input until EOF, then open a buffer with that text.
+Commands are read from standard error, which should be associated with a
```
I find it very offputting to spell out stdin and stderr. The *only* people for whom stdin/out/err is useful are people who will actually be *more confused* by the pronouncement "standard input". People who aren't used to these concepts won't be helped by the non-abbreviated form.

Pyrohh replied:
> The only people for whom stdin/out/err is useful are people who will actually be more confused by the pronouncement "standard input"

stdXXX are for programming contexts, where as standard XXX are not (this seems to be the trend in the majority of manual pages I looked at, GNU and BSD), as the former refers to the file descriptor. I don't know about others, but I don't see why people would be more confused: the non-abbreviated version is more correct in that it's a general descriptiong

Besides that, I don't see why we should deviate what seems to be the de facto standard in man page language.

justinmk replied:
`man cat` agrees with you. Meh.

justinmk commented on man/nvim.1:
```diff
@@ -0,0 +1,417 @@
+.It Fl
+Read text from standard input until EOF, then open a buffer with that text.
+Commands are read from standard error, which should be associated with a
+terminal.
```
I also think it's a hyper-correction to change this from the original wording, 

> Commands are read from stderr, which should be a tty.

The new form is totally meaningless to me. I don't know what "associated with a terminal" means, but I immediately understand "should be a tty". 

TTY is technical jargon with a specific meaning that is understood in context. No one in the context of a manpage mistakes it for physical teletypes.

Pyrohh replied:
How is it totally meaningless? Perhaps `assosciated` should be `refers to`, but besides that I don't understand how it's totally meaningless. If anything, TTY, besides being a misnomer, is less informative to users unaware of the term, where as "terminal" is self-explanatory. Someone might see "should be a terminal" and say "oh, well I'm running a terminal so I'm fine".

justinmk replied:
I don't get the insistence on TTY being a misnomer. Literally *everything* in computing is an analogy of some physical concept. TTY is a technical term with a long history in unix, as is "socket", "port", "core dump", "page", ...

It makes no sense to try to make it friendly, it won't help any user (who probably also doesn't know the difference between a terminal and a shell--not to mention a pseudo-terminal...).

> Someone might see "should be a terminal" and say "oh, well I'm running a terminal so I'm fine".

So they would then know how to send commands to `stderr`? That's quite optimistic :)

Pyrohh replied:
> I don't get the insistence on TTY being a misnomer. Literally everything in computing is an analogy of some physical concept. TTY is a technical term with a long history in unix, as is "socket", "port", "core dump", "page"

I hadn't thought about that, although It's still a misnomer in that it's usually not used in the context of Teleprinters, where are terminal makes more sense here because it's just a front-end to a system (like an interactive shell is to a login shell). It doesn't really matter much.

> It makes no sense to try to make it friendly, it won't help any user (who probably also doesn't know the difference between a terminal and a shell--not to mention a pseudo-terminal...).

I suppose I wasn't clear, but I'm more concerned about it being correct than friendly (not that they're mutually exclusive). And since "correct" isn't so easily defined in this context, I mean following the de facto standards regarding man pages.

The main point I'm making is that `tty` is to `terminal` as `stdXXX` is to `standard XXX`, at least from what I've gathered from various man pages. `tty` looks to be used almost exclusively in a non-general way, things like function names or `/dev` paths, which is no surprise given how UNIX influenced POSIX. However, even POSIX's [`man isatty`](http://pubs.opengroup.org/onlinepubs/009695399/functions/isatty.html) (among many others) uses the term "terminals", as there's no need to continue to use a misnomer just because the name is used. Most man pages I've read agree as well.

justinmk commented on man/nvim.1:
```diff
@@ -0,0 +1,417 @@
+command.
+See
+.Ic :help quickfix .
+.It There are a number of other options:
```
Can just say "Other options:"

Pyrohh replied:
I didn't do that because the above `-t` and `-q` are also options. Given that the DESCRIPTION has shrunk substantially, I just merged the two sections.

justinmk replied:
There really should be an options section. It's ok if description section contains the  options above this line.

Pyrohh replied:
I don't think we should add an options section for the sake of adding it, especially since it makes it inconsistent. It's pretty clear where the options start due to the indentation. The use of an options section is inconsistent among manuals, and doesn't really make sense here given how small the description is.

justinmk commented on man/nvim.1:
```diff
@@ -0,0 +1,417 @@
+See
+.Ic :help credits .
+.Pp
+Although hardly any of the original code remains,
```
can just delete this sentence.

Pyrohh replied:
:+1:

justinmk commented on man/nvim.1:
```diff
@@ -0,0 +1,417 @@
+command.
+.It Ev NVIM_TUI_CURSOR_SHAPE
+If set to
+.Cm 1 ,
```
non-zero

Pyrohh replied:
:+1:  `If defined` is good currently. I have a PR for that at #2569.

justinmk commented on man/nvim.1:
```diff
@@ -0,0 +1,417 @@
+Has no effect in GUIs.
+.It Ev NVIM_TUI_TRUE_COLOR
+If set to
+.Cm 1 ,
```
`If non-zero`

yagebu commented on man/nvim.1:
```diff
@@ -0,0 +1,411 @@
+.Ar pattern
+is omitted, the most recently used search pattern is used (if there is one).
+Note that
+.Cm /foo
```
`+/foo`

Pyrohh replied:
Nice catch, fixed.

yagebu commented on man/nvim.1:
```diff
@@ -0,0 +1,411 @@
+Up to 10 instances of these can be used, independently from instances of
+.Fl c .
+.It Fl c Ar command
+Execute
```
It should be explained that `+command` works the same way as `-c command`, so maybe add `+command` in the left column. From the lines above it might seem that the '+' only works with linenumbers and search patterns

Pyrohh replied:
Huh, good point. I really don't like how similar all these commands are...

Pyrohh replied:
Actually, I'm not sure... it's already like that in the existing documentation (although that's not much of a reason in itself, as we can just change the documentation).

At least for now, it might be easier to leave it, where it says that `+` should be used for line numbers, and the other ones (`+/`, `-c`, `--cmd`) should be used for other purposes, as opposed to saying something like `Here's four commands which are mostly the same but not really` (e.g., you can't do `+'e file'` but you can do `-c 'e file'`.

I think this is something that could be improved in the future, but I'm not sure about now. What do you think?

yagebu replied:
> it's already like that in the existing documentation

No ;) Actually `+command` is even listed first and it's also explained how you can use it with commands that include spaces (the '+' has to be between the quotes too):
  
    +{command}
    .TP
    \-c {command}
    {command} will be executed after the
    first file has been read.
    {command} is interpreted as an Ex command.
    If the {command} contains spaces it must be enclosed in double quotes (this
    depends on the shell that is used).
    Example: Vim "+set si" main.c

I definitely agree that it's not ideal to have all these very similar options, but as `+command` is the most general one and `+num` and `+/pat` are just special cases of it, I really think we should keep it.

yagebu replied:
> +num and +/pat are just special cases of it

To be exact, the one exception is that just a `+` will put the cursor at the end of the file.

Pyrohh replied:
Oops, well thanks for clearing that up. I'll update it and ping you after.

Commit:
Man pages: Move to nvim namespace and out of `runtime/`

Also clean up man page references in `runtime/doc/Makefile` as it
doesn't deal with them anymore.

Commit:
main.c: Make '-v' synonymous with '--version'

This is done for consistency with '-h' and '--help'.
While here, also move the mch_exit() call out of usage
to make it clear what is being done.

Helped-by: Justin M. Keyes <justinkz@gmail.com>

Commit:
`nvim --help` text: Unify with manpages

Commit:
English man pages: Convert to mdoc and rewrite for Nvim

Helped-by: Felipe Morales <hel.sheep@gmail.com>
Helped-by: Scott Prager <splinterofchaos@gmail.com>
Helped-by: Justin M. Keyes <justinkz@gmail.com>
Helped-by: Jakob Schnitzer <mail@jakobschnitzer.de>

Pyrohh commented:
@yagebu I attempted to modify the man page per your suggestion, what do you think?

Pyrohh commented:
It doesn't appear here, but the most recent suggestion was discussed here:
https://github.com/Pyrohh/neovim/commit/ba406b5c865a3c947b9a3fcec07764fee51c9245#commitcomment-11170474.

I'll merge after the tests.

Pyrohh commented:
One of the builds has stalled, but everything else passed so I'm merging this.
